hugo/content/playground/user-worker.ts

```diff
@@ -4,37 +4,77 @@
  * terms of the MIT License, which is available in the project root.
  ******************************************************************************/
 
-import { AstNode, DocumentState, createServicesForGrammar, startLanguageServer } from 'langium';
-import { createConnection } from 'vscode-languageserver/browser';
-import { PlaygroundWrapper, ByPassingMessageReader, ByPassingMessageWriter } from './monaco-utils';
-import { throttle } from './utils';
-
-const messageWrapper = new PlaygroundWrapper();
-const messageReader = new ByPassingMessageReader(self, messageWrapper);
-const messageWriter = new ByPassingMessageWriter(self, messageWrapper);
-
-const sendAst = throttle<AstNode>(1000, root => messageWriter.byPassWrite({
-    type: 'ast',
-    root
-}));
-
-messageReader.listenByPass(message => {
-    if(message.type === 'validated') {
-        sendAst.clear();
-        const connection = createConnection(messageReader, messageWriter);
-        createServicesForGrammar({
-            grammar: message.grammar, 
-            sharedModule: {
-                lsp: { Connection: () => connection}
+import { DocumentState, createServicesForGrammar, startLanguageServer } from 'langium';
+import { BrowserMessageReader, BrowserMessageWriter, createConnection, Diagnostic, Disposable, NotificationType } from 'vscode-languageserver/browser.js';
+
+// listen for messages to trigger starting the LS with a given grammar
+addEventListener('message', async (event) => {
+    if (event.data.type && event.data.type === 'startWithGrammar') {
+        if (event.data.grammar === undefined) {
+            throw new Error('User worker was started without a grammar!');
+        }
+        await startWithGrammar(event.data.grammar as string);
+    }
+});
+
+// disposable to remove previous build phase listeners
+let buildPhaseListener: Disposable;
+let newGrammarListener: Disposable;
+
+type DocumentChange = { uri: string, content: string, diagnostics: Diagnostic[] };
```

Comment:
`DocumentChange` & `DocumentChangeNotification`. Can they be moved into a common file? Saw them twice.

Reply:
Probably, I can look into that.

Reply:
It's been deduped now. Also noticed that there was some old logic that was able to be pruned, slimming down both workers even more 😎 .